master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala

```diff
@@ -653,11 +677,40 @@ private[celeborn] class Master(
       override def run(): Unit = {
         statusSystem.handleAppLost(appId, requestId)
         logInfo(s"Removed application $appId")
+        if (conf.hasHDFSStorage) {
+          cleanExpiredAppDirsOnHDFS(appId)
+        }
         context.reply(ApplicationLostResponse(StatusCode.SUCCESS))
       }
     })
   }
 
+  private def cleanExpiredAppDirsOnHDFS(dir: String = ""): Unit = {
+    if (hadoopFs == null) {
+      hadoopFs = CelebornHadoopUtils.getHadoopFS(conf)
+    }
+    val hdfsWorkPath = new Path(conf.hdfsDir, conf.workerWorkingDir)
+    if (hadoopFs.exists(hdfsWorkPath)) {
+      if (!dir.isEmpty) {
+        // delete specific app dir on application lost
+        hadoopFs.delete(new Path(hdfsWorkPath, dir), true)
+      } else {
+        val startTime = System.currentTimeMillis()
+        val cleanDirs = new mutable.ListBuffer[String]
+        val iter = hadoopFs.listStatusIterator(hdfsWorkPath)
+        while (iter.hasNext) {
+          val fileStatus = iter.next()
+          if (!statusSystem.appHeartbeatTime.containsKey(fileStatus.getPath.getName)) {
+            cleanDirs.append(fileStatus.getPath.toString)
+            hadoopFs.delete(fileStatus.getPath, true)
+          }
+        }
+        logInfo(
+          s"Clean all remnant HDFS dirs using ${System.currentTimeMillis() - startTime} ms, cleaned ${cleanDirs.mkString(",")}")
```

Comment:
```suggestion
        logInfo(
          s"Clean all remnant HDFS dirs using ${System.currentTimeMillis() - startTime} ms, cleaned ${cleanDirs.mkString(",")}")
```

log for each app dir, and use `Utils.msDurationToString` to format duration

Reply:
Updated. How about now?